genson/src/main/java/com/owlike/genson/convert/ClassMetadataConverter.java

```diff
@@ -101,4 +99,8 @@ public T deserialize(ObjectReader reader, Context ctx) throws Exception {
     }
     return wrapped.deserialize(reader, ctx);
   }
+
+  private boolean isJsonValueConverter(Converter<T> converter) {
+    return converter instanceof JSR353Bundle.JsonValueConverter;
```

Comment:
This would work only if the class metadata converter wraps the actual final converter, but this is not necessarily always true.

Maybe the right thing would be to:
   - check if the static type is a super type of the class resolved from metadata, if so then do use the resolved type
   - otherwise don't
   - eventually have an option to throw an exception if the type hierarchy mismatches?

Reply:
Good point... I wrote this before I fully understood how `ChainedFactory` works...

Technically, I probably don't need this check anymore, as I switched to using `JsonValueConverter` directly within `EvolvableHandler`, but I decided to leave it in because I do think that it should apply in a more general case: if I have `JsonValue`-typed attribute within a POJO, for example, whatever is read into it should be treated as an opaque value, without any annotation processing.

That said, I don't quite understand what you suggested above, and how type hierarchy would help us here. Can you please clarify? Again, the goal is to disable special `@class` annotation processing once the JSON-P barrier is crossed, and to treat it just like any other attribute.

Reply:
So your goal is to ensure that when you do `genson.deserialize("", JsonValue.class)` and class metadata is enabled, we don't deserialize to the actual @class value, right?
The `tClass` attribute of ClassMetadataConverter will be in this case `JsonValue`.

```
if (tClass.isAssignableFrom(classFromMetadata)) {
  genson.deserialize(..., classFromMetadata);
} else {
   wrapped.deserialize(reader, ctx);
}
```

So if we were doing something like the above, we would ensure that we respect the type hierarchy that the user did statically ask for (in this case JsonValue). Of course in some cases we can argue that if the type doesn't match we should throw an error, this could be the configurable part.

But anyway as you directly use the converter this is not needed.

Reply:
Right, that makes sense. I completely missed that `tClass` will be `JsonValue` or one of its subclasses.

That is much better thing to check than the converter chain, I'll make the change.